src/pygeoprocessing/geoprocessing.py

```diff
@@ -2494,37 +2486,49 @@ def convolve_2d(
             kernel_block[numpy.isclose(kernel_block, kernel_nodata)] = 0.0
         kernel_sum += numpy.sum(kernel_block)
 
+    work_queue = queue.Queue()
+    signal_offset_list = list(iterblocks(s_path_band, offset_only=True))
+    kernel_offset_list = list(iterblocks(k_path_band, offset_only=True))
+    n_blocks = len(signal_offset_list) * len(kernel_offset_list)
+
+    LOGGER.debug(f'start fill work queue thread')
+
+    def _fill_work_queue():
+        """Asynchronously fill the work queue."""
+        LOGGER.debug('fill work queue')
+        for signal_offset in signal_offset_list:
+            for kernel_offset in kernel_offset_list:
+                work_queue.put((signal_offset, kernel_offset))
+        work_queue.put(None)
+        LOGGER.debug('work queue full')
+
+    fill_work_queue_worker = threading.Thread(
+        target=_fill_work_queue)
+    fill_work_queue_worker.daemon = True
+    fill_work_queue_worker.start()
+
     # limit the size of the write queue so we don't accidentally load a whole
     # array into memory, work queue is okay because it's only passing block
     # indexes
-    work_queue = queue.Queue()
-    write_queue = queue.Queue(n_threads * 2)
-
-    # workers is n_threads - 1 because we count the current thread
+    LOGGER.debug(f'start worker thread')
+    write_queue = queue.Queue(10)
     worker_list = []
-    for worker_id in range(max(1, n_threads-1)):
-        worker = threading.Thread(
-            target=_convolve_2d_worker,
-            args=(
-                signal_path_band, kernel_path_band,
-                ignore_nodata_and_edges, normalize_kernel,
-                work_queue, write_queue))
-        worker.daemon = True
-        worker.start()
-        worker_list.append(worker)
-
-    n_blocks = 0
-    for signal_offset in iterblocks(s_path_band, offset_only=True):
-        for kernel_offset in iterblocks(k_path_band, offset_only=True):
-            work_queue.put((signal_offset, kernel_offset))
-            n_blocks += 1
-    for _ in range(max(1, n_threads-1)):
-        # signal end to worker
-        work_queue.put(None)
+    worker = threading.Thread(
+        target=_convolve_2d_worker,
+        args=(
+            signal_path_band, kernel_path_band,
+            ignore_nodata_and_edges, normalize_kernel,
+            set_tol_to_zero, work_queue, write_queue))
+    worker.daemon = True
+    worker.start()
 
     # used to count how many workers are still running
```

Comment:
is this comment needed anymore since `n_active_workers` is gone?

Reply:
Removed in 34802e5be006a4dbd7a9e0bb18330478cec6154a